pkg/policy/l4.go

```diff
@@ -358,12 +356,13 @@ func (l4 *L4Filter) IdentitySelectionUpdated(selector CachedSelector, selections
 	// that we could not push updates on a stale policy.
 	l4Policy := (*L4Policy)(atomic.LoadPointer(&l4.policy))
 	if l4Policy != nil {
+		derivedFrom := l4.DerivedFromRules.DeepCopy().Sort()
```

Comment:
Why do we need to:
1) DeepCopy()?
2) Sort the rules?

Can't both of these be avoided if we simply insert the labels by order when `l4.DerivedFromRules` is created / appended?

Reply:
They need to be sorted because the insert is not deterministic (see commit message) and we need to be able to have them sorted so we can compare them here:
https://github.com/cilium/cilium/blob/b50f082f2c4bf35d09d2d656eca204c85786fbb1/pkg/policy/mapstate.go#L114

They need to be copied to avoid accidentally sorting the original array. As an added bonus, sorting them also makes the unit tests much easier to read.

Reply:
But here we are doing a DeepCopy of the full array plus sorting it. Which operation will be performed more frequently? This one or when we insert a new label into `o.DerivedFromRules`?

Reply:
My understanding of the policy code is still somewhat limited, so I might be wrong. But it is my understanding that inserting a new label is much more frequent, as it is called for each rule of each policy applicable to the endpoint when the endpoint is regenerated. While this code here is only called once for each endpoint regeneration.

Reply:
If I'm not mistaken (/cc @jrajahalme) `IdentitySelectionUpdated()` can be invoked any time any new identity appears anywhere in the cluster. In large-scale clusters with constant churn, this may happen quite frequently.

Reply:
I see, that's a good pointer. That's a case then where we indeed should probably sort `L4Filter.DerivedFromRules` once we are certain that there are no rules which can be added to the `L4Filter` anymore. I will have to look a bit more deeply where that place would be.

Reply:
@joestringer is correct, typically policy computations are less frequent than map state generation due to identities being created either for new endpoints or newly resolved DNS names.

@gandro `L4Filter.attach()` is called once when the merging is over to "attach" the computed policy to the filter. Sorting in the beginning of this function is likely the right thing to do.

Reply:
Thank you all for the feedback, I think that helped me quite understand the code quite a bit better! I have pushed a new version where we sort the labels in `L4Filter.attach`. This also allows me to avoid having to create a copy, since this call seems to happen after any modifications to the `DerivedFromRules`.